### PR TITLE
Issue #59 Create UsersDAO

### LINKED ISSUE #59: Create UserDao

### PR DESCRIPTION
**Pull request by a-deriabin**

### CODE REVIEW

File: src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        }
+    }
+
+    @Override
+    public UsersEntity get(int id) {
+        DBConnectorInterface dbConnector = DBConnectorPostgres.getInstance();
+        try (Connection connection = dbConnector.getConnection()) {
+            Statement statement = connection.createStatement(
+                    ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY);
+            ResultSet resultSet = statement.executeQuery(GET_QUERY);
```
Comment (JustRevolt) [inline src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java]: How method will complete, if you do not set parameter in Query statement?

Comment (JustRevolt) [inline src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java]: How method will complete, if you do not set parameter in Query statement?

**Review by JustRevolt — APPROVED** [review]

File: src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java (hunk 1 above)
Comment (JustRevolt) [inline src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java]: How method will complete, if you do not set parameter in Query statement?

**Review by JustRevolt — CHANGES_REQUESTED** [review]

**Review by JustRevolt — APPROVED** [review]

**Review by nikaalin — APPROVED** [review]

**Review by ChirkinaTanya — APPROVED** [review]

File: src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
+package com.ifmo.epampractice.daoimpl;
+
+import com.ifmo.epampractice.dao.BasicDaoInterface;
```
Comment (PavelEzhkov) [inline src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java]: unused import

File: src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                    ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY);
+            ResultSet resultSet = statement.executeQuery(GET_ALL_QUERY);
+            while (resultSet.next()) {
+                users.add(entityFromResultSet(resultSet));
+            }
+            return users;
+        }
+        catch (SQLException e) {
+            e.printStackTrace();
+            //TODO: use logger when it will be set up.
```
Comment (PavelEzhkov) [inline src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java]: we have the logger

File: src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            return users;
```
Comment (PavelEzhkov) [inline src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java]: you can use one return for the method

File: src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        List<UsersEntity> users = new ArrayList<>();
+
+        DBConnectorInterface dbConnector = DBConnectorPostgres.getInstance();
+        try (Connection connection = dbConnector.getConnection()) {
+            Statement statement = connection.createStatement(
+                    ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY);
+            ResultSet resultSet = statement.executeQuery(GET_ALL_QUERY);
+            while (resultSet.next()) {
+                users.add(entityFromResultSet(resultSet));
+            }
```
Comment (PavelEzhkov) [inline src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java:47]: You have to close Statement and ResultSet as soon as you are finished with them or add  into try-with-resources

File: src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    private static final String DELETE_QUERY = "DELETE FROM Users WHERE ID = ?";
+    private static final String INSERT_QUERY = "INSERT INTO Users (Email, Password_hash, Role, Name, Birth_date, " +
+            "Signup_date, Pass_id, Driving_license_id, Contact_phone, Address, Is_blocked, " +
+            "Reputation) VALUES (?, ?, ?::e_role_users, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
+
+    @Override
+    public List<UsersEntity> getAll() {
+        List<UsersEntity> users = new ArrayList<>();
+
+        DBConnectorInterface dbConnector = DBConnectorPostgres.getInstance();
```
Comment (PavelEzhkov) [inline src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java]: You've wrote it  in each method. You can do it 1 time

File: src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            statement.execute();
+        }
+        catch (SQLException e) {
+            e.printStackTrace();
+            //TODO: use logger when it will be set up.
+        }
+    }
+
+    @Override
+    public void delete(UsersEntity entity) {
```
Comment (PavelEzhkov) [inline src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java:88]: Did you try the method? 
`org.postgresql.util.PSQLException: ERROR: update or delete on table "users" violates foreign key constraint "orders_client_id_fkey" on table "orders"
  Detail: Key (id)=(1) is still referenced from table "orders".`

probably I have to update my db

File: src/main/java/com/ifmo/epampractice/dao/UsersDao.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+import com.ifmo.epampractice.entity.UsersEntity;
+
+import java.util.List;
+
+public interface UsersDao extends BasicDaoInterface<UsersEntity> {
+    List<UsersEntity> getAll();
+
+    UsersEntity get(int id);
+
+    void update(UsersEntity o);
```
Comment (PavelEzhkov) [inline src/main/java/com/ifmo/epampractice/dao/UsersDao.java]: Why "o"? userEntity will be better

File: src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+    @Override
+    public UsersEntity get(int id) {
+        DBConnectorInterface dbConnector = DBConnectorPostgres.getInstance();
+        try (Connection connection = dbConnector.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(GET_QUERY,
+                    ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY);
+            statement.setInt(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (!resultSet.next()) {
```
Comment (PavelEzhkov) [inline src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java]: is not the best way, the thing is that `!resultSet.next()` makes reading more confusing than just `resultSet.next()`
try:
```
            if (resultSet.next()) {
                return entityFromResultSet(resultSet);
            }
        } catch (SQLException e) {
          ...
        }
        return null;
```

**Review by PavelEzhkov — COMMENTED** [review]
You can do it better and merge dev, please

**Review by Nursatb — APPROVED** [review]

File: src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java (hunk 7 above)
Comment (a-deriabin) [reply src/main/java/com/ifmo/epampractice/daoimpl/UsersDaoImpl.java:88]: The problem is in DDL code.
Created Issue #77 for this.

**Pull request merged**